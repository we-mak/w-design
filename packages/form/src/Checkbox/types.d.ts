import {
  GlobProps,
  InputControlProps,
  SyntheticEventProps,
  MouseEventProps
} from "../../../types/typings";

export interface CheckboxProps
  extends GlobProps,
    InputControlProps,
    SyntheticEventProps,
    MouseEventProps {
  //
  checkboxRef?: React.RefAttributes;
  name?: string;
  defaultChecked?: boolean;
  isChecked?: boolean;
  isDisabled?: boolean;
  isRequired?: boolean;
  isIndeterminate?: boolean;
  label?: string;
  value?: number | string;
  ariaRequired?: boolean;
}
