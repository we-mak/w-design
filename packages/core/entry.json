{
  "index": "src/index.ts",
  "AnchorScroll": "src/AnchorScroll/index.tsx",
  "Avatar": "src/Avatar/index.tsx",
  "AvatarGroup": "src/AvatarGroup/index.tsx",
  "Box": "src/Box/index.ts",
  "Breadcrumbs": "src/Breadcrumbs/index.tsx",
  "Button": "src/Button/index.tsx",
  "Card": "src/Card/index.tsx",
  "Comment": "src/Comment/index.tsx",
  "Container": "src/Container/index.tsx",
  "CustomTag": "src/CustomTag/index.tsx",
  "DisplayElement": "src/DisplayElement/index.tsx",
  "Divider": "src/Divider/index.ts",
  "Drawer": "src/Drawer/index.tsx",
  "Dropdown": "src/Dropdown/index.tsx",
  "EmptyView": "src/EmptyView/index.tsx",
  "Flexbox": "src/Flexbox/index.tsx",
  "Icon": "src/Icon/index.tsx",
  "Image": "src/Image/index.tsx",
  "List": "src/List/index.tsx",
  "Menu": "src/Menu/index.tsx",
  "Modal": "src/Modal/index.tsx",
  "Navbar": "src/Navbar/index.tsx",
  "Picture": "src/Picture/index.tsx",
  "Portal": "src/Portal/index.tsx",
  "ProgressLoader": "src/ProgressLoader/index.tsx",
  "Provider": "src/Provider/index.tsx",
  "PushMessage": "src/PushMessage/index.tsx",
  "SectionMessage": "src/SectionMessage/index.tsx",
  "Spinner": "src/Spinner/index.tsx",
  "Typo": "src/Typo/index.tsx"
}
