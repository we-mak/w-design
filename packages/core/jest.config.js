const base = require("../../jest.config.base.js");
const pack = require("./package.json");

module.exports = {
  ...base,
  name: pack.name,
  displayName: pack.name,
  rootDir: "../..",
  testMatch: [
    `<rootDir>/packages/core/**/*.test.+(ts|tsx|js)`,
    `<rootDir>/packages/core/**/__tests__/*.+(ts|tsx|js)`
  ]
};
