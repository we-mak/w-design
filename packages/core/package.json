{
  "name": "@w-design/core",
  "version": "0.41.0-alpha.28+d990095",
  "description": "w-design core components",
  "keywords": [
    "w-design",
    "react",
    "core",
    "components"
  ],
  "author": "Son_dang <son.dang@doda100.com >",
  "homepage": "https://github.com/we-mak/w-design#readme",
  "license": "MIT",
  "main": "lib/cjs/index.js",
  "module": "lib/es/index.es.js",
  "types": "lib/cjs/index.d.ts",
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib/cjs/*",
    "lib/es/*"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://viiiprock@github.com/we-mak/w-design.git"
  },
  "scripts": {
    "test": "cross-env CI=1 jest test -u --colors"
  },
  "bugs": {
    "url": "https://github.com/we-mak/w-design/issues"
  },
  "dependencies": {
    "@types/body-scroll-lock": "^2.6.1",
    "@w-design/helpers": "^0.40.0",
    "@w-design/hooks": "^0.40.0",
    "body-scroll-lock": "^2.6.4"
  },
  "gitHead": "d990095e2f75e0d6e34aff66a98969a4de8eae6f"
}
