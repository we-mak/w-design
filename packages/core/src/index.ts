export { default as AnchorScroll } from "./AnchorScroll";
export { default as Avatar } from "./Avatar";
export { default as AvatarGroup } from "./AvatarGroup";
export { default as Badge } from "./Badge";
export { default as Box } from "./Box";
export { default as Breadcrumbs, BreadcrumbsItem } from "./Breadcrumbs";
export { default as Button, ButtonGroup } from "./Button";
export {
  default as Card,
  CardMedia,
  CardHeader,
  CardContent,
  CardFooter,
} from "./Card";

export * from "./Chip";
export { default as Chip } from "./Chip";

export { default as Collapse } from "./Collapse";
export { default as Comment } from "./Comment";
export { default as Container } from "./Container";
export { default as CustomTag } from "./CustomTag";
export { default as DisplayElement } from "./DisplayElement";
export { default as Divider } from "./Divider";
export { default as Drawer } from "./Drawer";
export { default as Dropdown } from "./Dropdown";
export { default as EmptyView } from "./EmptyView";
export { default as Flexbox } from "./Flexbox";
export { default as Icon } from "./Icon";
export { default as Image, ImageRatio } from "./Image";
export { default as List, ListItem, ListFooter } from "./List";
export { default as Menu, MenuItem, MenuHeading, SubMenu } from "./Menu";
export { default as Modal } from "./Modal";
export { default as Navbar } from "./Navbar";

export * from "./Pagination";
export { default as Pagination } from "./Pagination";

export { default as Picture } from "./Picture";
export { default as Portal } from "./Portal";

export * from "./Popover";
export { default as Popover } from "./Popover";

export { default as ProgressLoader } from "./ProgressLoader";
export { default as Provider } from "./Provider";
export { default as SectionMessage } from "./SectionMessage";

export * from "./Spinner";
export { default as Spinner } from "./Spinner";

export * from "./Steps";
export { default as Steps } from "./Steps";

export { default as Tabs, Tab } from "./Tabs";
export { default as Typo } from "./Typo";

export * from "./PushMessage";
export { default as PushMessage } from "./PushMessage";
