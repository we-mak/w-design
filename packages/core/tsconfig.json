{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "lib",
    "declarationDir": "lib",
    "baseUrl": "."
  },
  "include": ["src/**/*"]
}
