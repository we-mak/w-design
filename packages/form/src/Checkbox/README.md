| Props          | type                                             | default | description             |
| -------------- | ------------------------------------------------ | ------- | ----------------------- |
| CheckboxRef    | React.MutableRefObject                           |         | checkbox node reference |
| name           | string                                           |         |                         |
| defaultChecked | boolean                                          | false   |                         |
| isChecked      |                                                  | false   |                         |
| isDisabled     |                                                  | false   |                         |
| isRequired     |                                                  | false   |                         |
| label          |                                                  |         |                         |
| value          | number, string                                   |         |                         |
| className      |                                                  |         |                         |
| id             |                                                  |         |                         |
| onChange       | (e: React.ChangeEvent<HTMLInputElement>) => void |         |                         |
| onInput        | (e: React.FormEvent<HTMLInputElement>) => void   |         |                         |
