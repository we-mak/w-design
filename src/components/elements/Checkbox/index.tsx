import * as React from "react";
import styled, { css } from "styled-components";
import FormGroup from "../../components/FormGroup";
import { CheckboxProps } from "./types";

const Label = styled.label<CheckboxProps>`
  display: flex;
  align-items: flex-start;
  position: relative;
  line-height: 1.2;
  margin: ${props => props.theme.space[1]}px 0;
  user-select: none;
  &:hover {
    cursor: ${props => (props.isDisabled ? `not-allowed` : `pointer`)};
  }

  ${props => props.isDisabled && `opacity: 0.7`};
`;
Label.displayName = "Label";

const LabelText = styled.span`
  padding: 5px 0;
`;
LabelText.displayName = "LabelText";

const CheckboxInput = styled.input`
  clip: rect(0, 0, 0, 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
`;
CheckboxInput.displayName = "CheckboxInput";

// Get checked style for the icon
const getCheckedStyle = (props: CheckboxProps) => {
  if (props.isChecked) {
    return css`
      color: ${props.theme.colors["B50"]};
      fill: ${props.theme.colors["N1"]};
      rect {
        &:first-child {
          stroke: ${props.theme.colors["B50"]};
        }
      }
    `;
  }

  return css`
    color: ${props.theme.colors["N1"]};
    fill: transparent;
    rect {
      &:first-child {
        stroke: ${props.theme.colors["N5"]};
      }
    }
  `;
};

const CheckboxIcon = styled.span`
  line-height: 0;
  flex-shrink: 0;
  cursor: pointer;
  rect {
    &:first-child {
      stroke-width: 0.08rem;
      transition: stroke 0.2s ease-in-out 0s;
    }
  }
  ${getCheckedStyle}
`;
CheckboxIcon.displayName = "CheckboxIcon";

// TODO: nested checkboxes
const Checkbox = ({
  checkboxRef,
  name,
  label,
  defaultChecked = false,
  isDisabled = false,
  isRequired,
  value,
  onChange,
  ...rest
}: CheckboxProps) => {
  const [isChecked, setChecked] = React.useState(defaultChecked);

  const onCheckboxChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (isDisabled) return null;
    e.persist();
    // set onChange props
    if (onChange) onChange(e);

    if (e.target.checked !== undefined) {
      setChecked(e.target.checked);
    }

    return true;
  };

  return (
    <FormGroup>
      <Label isDisabled={isDisabled}>
        <CheckboxInput
          ref={checkboxRef}
          type="checkbox"
          name={name}
          checked={isChecked}
          disabled={isDisabled}
          value={value}
          onChange={onCheckboxChange}
          required={isRequired}
          ariaRequired={isRequired}
          {...rest}
        />
        <CheckboxIcon isChecked={isChecked}>
          <svg width="28" height="28" viewBox="0 0 24 24" focusable="false" role="presentation">
            <g fillRule="evenodd">
              <rect fill="currentColor" x="6" y="6" width="12" height="12" rx="2" />
              <path
                d="M9.707 11.293a1 1 0 1 0-1.414 1.414l2 2a1 1 0 0 0 1.414 0l4-4a1 1 0 1 0-1.414-1.414L11 12.586l-1.293-1.293z"
                fill="inherit"
              />
            </g>
          </svg>
        </CheckboxIcon>
        <LabelText>{label}</LabelText>
      </Label>
    </FormGroup>
  );
};

export default Checkbox;
