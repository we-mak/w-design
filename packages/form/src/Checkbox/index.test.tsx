import * as React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { Provider } from "@w-design/core";
import Checkbox from "./index";

afterEach(cleanup);

describe("<Checkbox/>", () => {
  // Render test
  it("should renders correctly", () => {
    const { container } = render(
      <Provider>
        <Checkbox name="Checkbox" label="Test checkbox" />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  // Props
  it("should checked by default", () => {
    const { getByTestId } = render(
      <Provider>
        <Checkbox defaultChecked />
      </Provider>
    );
    const checkbox = getByTestId("checkbox");
    expect((checkbox as HTMLInputElement).checked).toEqual(true);
  });

  // Event test
  it("should fire check event", () => {
    const { getByTestId } = render(
      <Provider>
        <Checkbox />
      </Provider>
    );
    const checkbox = getByTestId("checkbox");
    expect((checkbox as HTMLInputElement).checked).toEqual(false);
    fireEvent.click(checkbox);
    expect((checkbox as HTMLInputElement).checked).toEqual(true);
    fireEvent.click(checkbox);
    expect((checkbox as HTMLInputElement).checked).toEqual(false);
  });

  // Others
  it("shouldn't be changed if isDisabled", () => {
    const { getByTestId } = render(
      <Provider>
        <Checkbox isDisabled />
      </Provider>
    );
    const disabledCheckbox = getByTestId("checkbox");
    expect((disabledCheckbox as HTMLInputElement).checked).toEqual(false);
    fireEvent.click(disabledCheckbox);
    expect((disabledCheckbox as HTMLInputElement).checked).toEqual(false);
  });

  it('calls "onChange" prop on change', () => {
    const onChange = jest.fn();
    const { getByTestId } = render(
      <Provider>
        <Checkbox onChange={onChange} />
      </Provider>
    );
    fireEvent.click(getByTestId("checkbox"));
    expect(onChange).toHaveBeenCalled();
  });
});
