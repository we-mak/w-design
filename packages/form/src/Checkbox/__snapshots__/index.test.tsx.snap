// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox/> should renders correctly 1`] = `
<div>
  <div
    class="sc-bdnylx gXXlVO"
  >
    <label
      class="sc-gtssRu hoCjmA"
    >
      <input
        class="sc-hKFyIo bJEiep"
        data-testid="checkbox"
        name="Checkbox"
        type="checkbox"
        value=""
      />
      <span
        class="sc-eCApGN cLUYyx"
      >
        <svg
          focusable="false"
          height="28"
          role="presentation"
          viewBox="0 0 24 24"
          width="28"
        >
          <g
            fill-rule="evenodd"
          >
            <rect
              fill="currentColor"
              height="12"
              rx="2"
              width="12"
              x="6"
              y="6"
            />
            <path
              d="M9.707 11.293a1 1 0 1 0-1.414 1.414l2 2a1 1 0 0 0 1.414 0l4-4a1 1 0 1 0-1.414-1.414L11 12.586l-1.293-1.293z"
              fill="inherit"
            />
          </g>
        </svg>
      </span>
      <span
        class="sc-dlnjPT lpjboc"
      >
        Test checkbox
      </span>
    </label>
  </div>
</div>
`;
